appservice/src/createAppService/SiteRuntimeStep.ts

```diff
@@ -79,19 +79,36 @@ export class SiteRuntimeStep extends AzureWizardPromptStep<IAppServiceWizardCont
         const runtimesParsed: ApplicationStackJsonResponse = <ApplicationStackJsonResponse>JSON.parse(runtimes);
 
         return runtimesParsed.value.map((runtime) => {
-            return nonNullProp(runtime.properties, 'majorVersions').map((majorVersion) => {
+            return nonNullProp(runtime.properties, 'majorVersions').sort((rt1, rt2) => {
+                if (rt1.isDefault) {
+                    return -1;
+                } else if (rt2.isDefault) {
+                    return 1;
+                } else {
+                    // runtimVersion comes in the format RUNTIME|X.X
+                    const runtimeRegExp: RegExp = /.*(?=\|)+./g;
+                    const v1: number = parseFloat(nonNullProp(rt1, 'runtimeVersion').replace(runtimeRegExp, ''));
+                    const v2: number = parseFloat(nonNullProp(rt2, 'runtimeVersion').replace(runtimeRegExp, ''));
+                    return v2 - v1;
+                }
+            }).map((majorVersion) => {
                 return { name: majorVersion.runtimeVersion, display: majorVersion.displayVersion, isDefault: majorVersion.isDefault };
             });
         }).reduce((acc, val) => acc.concat(val));
         // this is to flatten the runtimes to one array
     }
 
-    private sortQuickPicksByRuntime(runtimeItems: IAzureQuickPickItem<ApplicationStack>[], recommendedRuntimes: string[]): IAzureQuickPickItem<ApplicationStack>[] {
-        function getPriority(item: IAzureQuickPickItem<ApplicationStack>): number {
-
-            const index: number = recommendedRuntimes.findIndex((runtime: string) => nonNullProp(item.data, 'name').includes(runtime));
-            return index === -1 ? recommendedRuntimes.length : index;
+    private sortQuickPicksByRuntime(runtimeQuickPicks: IAzureQuickPickItem<ApplicationStack>[], runtimeRecommendations: string[]): IAzureQuickPickItem<ApplicationStack>[] {
+        const recommendedQuickPicks: IAzureQuickPickItem<ApplicationStack>[] = [];
+        // go backwards so that when runtime is unshifted to front, the order is unaffected
+        for (let i: number = runtimeQuickPicks.length - 1; i >= 0; i = i - 1) {
```

Comment:
So this method is called `sortQuickPicksByRuntime`. I guess I'd prefer if all the sorting logic was in this one function rather than having half of it in `getLinuxRuntimeStack`. Also I prefer using the `.sort` method instead of this for/unshift/concat logic because it makes me nervous about an off-by-one error

Reply:
I think I'd rather rename this to `shiftRecommendedToTop` as there is no actual sorting going on here.

I don't want to do actual sorting here because it's a lot easier to sort by runtime before I've flattened the array (when I make the API call).  

Since we're traversing the array backward, I think even if we splice out an element, we shouldn't have any indexing errors, but I can find another method.

Maybe I wasn't understanding the `sort` method correctly, but when it was moving recommended runtimes to the top, it was swapping the recommended runtime's index with whatever was previous at the top which is why sometimes the versions would be out of order or would get sandwiched in between other runtimes.

I was checking if either item was the recommended runtime and returned -1 or 1 appropriately, but if not, returned 0.  That should have caused it to keep its position, but since it's getting swapped with the recommended runtime, that position was incorrect.

Reply:
I'm more okay with this if you change the name. However in an ideal world this all has unit-tests and that would be easier to do if it was one function.

You can definitely do all of this inside one sort function. I don't think the behavior is guaranteed when you return 0, which is why it would have weird results. I think this sort logic would work:
```
const aInfo = getRuntimeInfo(a);
const bInfo = getRuntimeInfo(b);
if (aInfo.name !== bInfo.name) {
    return getPriority(a) - getPriority(b);
} else if (aInfo.major !== bInfo.major) {
    return aInfo.major - bInfo.major;
} else {
    return aInfo.minor - bInfo.minor;
}

function getRuntimeInfo() {
// use a regexp to get name, major, and minor from node|8.0
}
```